src/common/attribute.rs

```diff
@@ -0,0 +1,626 @@
+		colormap: Option<CopyableFromParent<Colormap>>,
+		cursor_appearance: Option<Option<CursorAppearance>>,
+	) -> Self {
+		let mut x11_size = 0;
+		let mut mask = AttributeMask::empty();
+
+		if let Some(background_pixmap) = &background_pixmap {
+			x11_size += background_pixmap.x11_size();
+			mask |= AttributeMask::BACKGROUND_PIXMAP;
+		}
+		if let Some(background_pixel) = &background_pixel {
+			x11_size += background_pixel.x11_size();
+			mask |= AttributeMask::BACKGROUND_PIXEL;
+		}
+
+		if let Some(border_pixmap) = &border_pixmap {
+			x11_size += border_pixmap.x11_size();
+			mask |= AttributeMask::BORDER_PIXMAP;
+		}
+		if let Some(border_pixel) = &border_pixel {
+			x11_size += border_pixel.x11_size();
+			mask |= AttributeMask::BORDER_PIXEL;
+		}
+
+		if let Some(bit_gravity) = &bit_gravity {
+			x11_size += bit_gravity.x11_size();
+			mask |= AttributeMask::BIT_GRAVITY;
+		}
+		if let Some(win_gravity) = &win_gravity {
+			x11_size += win_gravity.x11_size();
+			mask |= AttributeMask::WIN_GRAVITY;
+		}
+
+		if let Some(backing_store) = &backing_store {
+			x11_size += backing_store.x11_size();
+			mask |= AttributeMask::BACKING_STORE;
+		}
+		if let Some(backing_planes) = &backing_planes {
+			x11_size += backing_planes.x11_size();
+			mask |= AttributeMask::BACKING_PLANES;
+		}
+		if let Some(backing_pixel) = &backing_pixel {
+			x11_size += backing_pixel.x11_size();
+			mask |= AttributeMask::BACKING_PIXEL;
+		}
+
+		if let Some(override_redirect) = &override_redirect {
+			x11_size += override_redirect.x11_size();
+			mask |= AttributeMask::OVERRIDE_REDIRECT;
+		}
+		if let Some(save_under) = &save_under {
+			x11_size += save_under.x11_size();
+			mask |= AttributeMask::SAVE_UNDER;
+		}
+
+		if let Some(event_mask) = &event_mask {
+			x11_size += event_mask.x11_size();
+			mask |= AttributeMask::EVENT_MASK;
+		}
+		if let Some(do_not_propagate_mask) = &do_not_propagate_mask {
+			x11_size += do_not_propagate_mask.x11_size();
+			mask |= AttributeMask::DO_NOT_PROPAGATE_MASK;
+		}
+
+		if let Some(colormap) = &colormap {
+			x11_size += colormap.x11_size();
+			mask |= AttributeMask::COLORMAP;
+		}
+
+		if let Some(cursor_appearance) = &cursor_appearance {
+			x11_size += cursor_appearance.x11_size();
+			mask |= AttributeMask::CURSOR_APPEARANCE;
+		}
+
+		Self {
+			x11_size,
+			mask,
+
+			background_pixmap,
+			background_pixel,
+
+			border_pixmap,
+			border_pixel,
+
+			bit_gravity,
+			win_gravity,
+
+			backing_store,
+			backing_planes,
+			backing_pixel,
+
+			// These booleans are converted into our [`Bool`] type so that they can easily be
+			// written as four bytes.
+			override_redirect: override_redirect.map(std::convert::Into::into),
+			save_under: save_under.map(Into::into),
+
+			event_mask,
+			do_not_propagate_mask,
+
+			colormap,
+
+			cursor_appearance,
+		}
+	}
+
+	#[must_use]
+	pub const fn background_pixmap(&self) -> &Option<ParentRelatable<Option<Pixmap>>> {
```

Comment:
Maybe there should be type aliases for these wrapper types for the fields?  Might be easier to read overall:
```rust
pub type BackgroundPixmapAttribute = ParentRelatable<Option<Pixmap>>;
pub type CursorAppearanceAttribute = Option<CursorAppearance>;
// etc.

pub const fn background_pixmap(&self) -> &Option<BackgroundPixmapAttribute> { ... }
pub const fn cursor_appearance(&self) -> &Option<CursorAppearanceAttribute>;
// etc.
```
I think the main readability advantage here is it clearly communicated what part of the type signature is indicating "this attribute is or isn't present" (the outermost `Option`), and which part matches the actual type of the attribute (`BackgroundPixmapAttribute` or `CursorAppearanceAttribute`, etc.).